Add bookings

kmkmaslov opened the pull request.

kmkmaslov committed:
initial

kmkmaslov committed:
fix pom

kmkmaslov committed:
refactor, add db, state, booking, comments

kmkmaslov committed:
pix pom? add db

kmkmaslov committed:
pix pom?

kmkmaslov committed:
add booking

kmkmaslov committed:
fix

kmkmaslov committed:
fix

kmkmaslov committed:
fix

kmkmaslov commented:
не получается подцепить файл application.properties

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.List;
+
+@Slf4j
+@RestController
```
@RestController
Поскольку далее Вы везде используете ResponseEntity, то здесь @RestController стоит заменить на @Controller, так как ResponseEntity уже оборачивает ответ в REST
https://coderlessons.com/articles/java/raznitsa-mezhdu-restcontroller-i-annotatsiei-controller-v-spring-mvc-i-rest

kmkmaslov replied:
выполнено

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -1,21 +1,41 @@
 
+@Entity
+@Table(name = "bookings")
 @Data
```
Data
Лучше не использовать эту аннотацию для сущностей. Дело в том, что ломбок генерирует EqualsAndHashCode, И могут быть конфликты с аналогичными механизмами hibernate’a
В данной работе ничего страшного не произойдет, но лучше не привыкать).
Если бы в проект добавили кэширование, то возникли бы проблемы.
https://stackoverflow.com/questions/75181366/why-jpa-buddy-complains-about-data-annotation-over-jpa-entity

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/repo/BookingRepository.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.List;
+import java.util.Optional;
+
+@Repository("BookingRepository")
```
Аннотация @Repository не требуется, так как наследуясь от JpaRepository - мы уже объявляем бин.
Здесь и в других интерфейсах. Посмотрите, пожалуйста.

kmkmaslov replied:
Выполнено

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,281 @@
+    private UserMapper userMapper;
+
+    @Override
+    public BookingDto addNewBooking(long userId, BookingFromUserDto bookingFromUser) {
```
Transaction

Над методами сервиса, которые меняют данные в БД (раз  используется data-jpa) нужна аннотация Transaction
Два варианта.
Если аннотация объявлена над классом, то для методов, меняющих данные в базе, её повторно объявлять не нужно, А а вот для методов выборки лучше добавить параметр readonly=true
Если аннотация на классом не объявлена, то её нужно объявить для всех методов, которые изменяют данные в базе, а для методов Выборки данных она вообще не нужна.

kmkmaslov replied:
Выполено

avfyodorov commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,281 @@
+    }
+
+    @Override
+    public BookingDto updateBooking(long userId, long bookingId, String approved) throws RuntimeException {
```
Transaction
И здесь и в других аналогичных местах проекта.

kmkmaslov replied:
Выполннео

avfyodorov commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,16 +1,52 @@
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof Item)) return false;
```
Лучше бы, конечно, избавиться от этих двух методов. Почти наверняка они не нужны. Ну, если всё-таки без них никак, то лучше изменить как-то так:

if (o == null || Hibernate.getClass(this) != Hibernate.getClass(o)) return false;

Дело в том, что у гибернейта свой механизм и могут быть конфликты. Это касается только классов-моделей, которые соответствуют таблицам в базе данных.

kmkmaslov replied:
Выполено

avfyodorov commented on src/main/resources/appication.yml:
```diff
@@ -0,0 +1,14 @@
+spring:
```
Действительно ли нужен этот файл в проекте? 
Разве одного application.properties не хватает?

kmkmaslov replied:
выполнено

avfyodorov commented on src/test/java/ru/practicum/shareit/ShareItTests.java:
```diff
@@ -1,13 +1,14 @@
 package ru.practicum.shareit;
 
-import org.junit.jupiter.api.Test;
+/*import org.junit.jupiter.api.Test;
```
Тогда лучше совсем удалить этот класс из проекта.

kmkmaslov replied:
Выполнено

avfyodorov reviewed (commented):
Добрый вечер, Константин! 
В целом хорошо, Правда, есть ряд уточнений. 
Нужно ли  одновременно два файла настроек в каталоге ресурсов? 
Ещё один момент-нужно добиться прохождение тестов на гитхабе, т.н. Зелёная сборка.

kmkmaslov committed:
corrections after review

kmkmaslov committed:
corrections 2nd after review

avfyodorov reviewed (commented):
Добрый вечер, Константин! 
Мои комментарии учтены, Замечаний нет. 
Ещё один момент - тестовая сборка на github'e обязательно должна быть зелёная.  Если у Вас возникли сложности, пожалуйста, обратитесь к наставнику.

kmkmaslov committed:
Corrected comments from the mentor

kmkmaslov committed:
Corrected comments from the mentor

kmkmaslov committed:
fix LocalTime

kmkmaslov committed:
Sergei Chebotarev

05:58
Попробуй использовать аннотацию @RestController  вместо @Controller
и @RestControllerAdvice вместо @ControllerAdvice
Controller не работает обработка ошибок нормально, поэтому ты вернул по моему совету

kmkmaslov committed:
add ExceptionHandler HttpStatus.CONFLICT

kmkmaslov committed:
fix ExceptionHandler; add annotation

kmkmaslov committed:
fix  @PathVariable annotation

kmkmaslov committed:
fix  @RequestParam annotation

kmkmaslov committed:
fix  @PathVariable

kmkmaslov committed:
fix  @PathVariable

kmkmaslov committed:
fix  config

kmkmaslov committed:
fix  variable
